File: python/src/dynmagma.py
from typing import Callable, Any, TypeVar, Iterable, Optional, Tuple


class DynMagma (Magma):
    def __init__(self,
                 gen1: Callable[[], Iterator[Any]],
                 op1: Callable[[Any, Any], Any],
                 member1: Callable[[Any], bool]):
        self.gen1 = gen1
        self.op1 = op1
        self.member1 = member1

    def __repr__(self):
        return "dyn"

    def gen(self) -> Iterator[Any]:
        for i in self.gen1():
            yield i

    def op(self, a, b) -> Any:
        return self.op1(a, b)

    def member(self, a) -> HeavyBool:
        if self.member1(a):
            return HeavyTrue(f"{a} is a member")
        else:
            return HeavyFalse(f"{a} is not a member")

File: scala/src/main/scala/heavybool/examples/DynMagma.scala
package heavybool.examples

import cats.Foldable
import heavybool.HeavyBool.{HFalse, HTrue}
import heavybool.{HeavyBool, Magma}

case class DynMagma[T, C[_]:Foldable](gen1: () => C[T],
                       op1: (T, T) => T,
                       member1: T => Boolean) extends Magma[T, C] {
  override def toString: String = "dyn"

  def gen(): C[T] = gen1()

  def op(a: T, b: T): T = op1(a, b)

  def member(a: T): HeavyBool = member1(a) match {
    case true => HTrue ++ Map("reason" -> "a is a member",
                              "a" -> a)

    case false => HFalse ++ Map("reason" -> "a is not a member",
                                "a" -> a)
  }
}

